충남대 BE_이은경 2주차 과제(1단계)

pkyung opened the pull request:
## 구현 과정
1. ValidProductName 어노테이션을 생성하여 각각의 경우에 따라 다른 에러 메시지를 출력하도록 하였습니다.  -> 기본 validate 어노테이션을 활용하도록 리펙토링
2. controller에서 해당 에러를 ExceptionHandler로 받아 처리했습니다.  -> controller의 ExceptionHandler를 ProductExceptionHandler 클래스를 만들어서 모아서 처리하도록 리펙토링

## 과제 진행 사항
### 유효성 검사 및 예외 사항
- [x] 상품 이름은 공백 포함 15자까지 입력
- [x]  ( ), [ ], +, -, &, /, _ 외의 특수 문자 불가
- [x] '카카오' 가 포함된 문구는 담당 MD와 협의

pkyung committed:
Chore: 1단계 코드 이동

pkyung committed:
Chore: 기능 목록 이동

pkyung committed:
Chore: 1주차 피드백 반영 코드 이동

pkyung committed:
Docs: 기능 목록 작성 - 1단계

pkyung committed:
Refactor: final 키워드를 추가하여 autowired 제거

pkyung committed:
Docs: 기능 목록 수정

pkyung committed:
Feat: ValidProductName 어노테이션 생성

ValidProductName 어노테이션을 생성하고 ProductNameValidator를 구현하여
상품 이름의 길이가 15자 이내인지, 특정 특수 문자만 들어있는지, 카카오
라는 단어가 들어있는 지 확인하고, 그에 맞는 응답을 하도록 구현했다.
controller에 ExceptionHandler를 통해 validated로 발생되는
MethodArgumentNotValidException 에러를 잡아서 내부의 defaultMessage를
가져와 400 bad request 와 함꼐 메시지를 반환하도록 구현했다.

pkyung committed:
Refactor: 필요 없는 import 제거 및 코드 정렬

pkyung committed:
Docs: 기능 목록 구현 - 1단계

pkyung committed:
Merge branch 'pkyung' into main

ksy90101 commented on src/main/java/gift/validator/ProductNameValidator.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public boolean isValid(String name, ConstraintValidatorContext context) {
+        if (name == null || name.equals("")) {
+            return false;
+        }
```
이부분은 @NotEmpty, @NotNull과 같은 기본 validate로 해결해볼수 있지 않을까요?

ksy90101 commented on src/main/java/gift/validator/ProductNameValidator.java:
```diff
@@ -0,0 +1,42 @@
+            return false;
+        }
+
+        if (name.length() > 15) {
+            context.disableDefaultConstraintViolation();
+            context.buildConstraintViolationWithTemplate("상품 이름은 15자 이내여야 합니다.")
+                    .addConstraintViolation();
+            return false;
```
이 부분은 @Length로 해결해볼수 있지 않을까요?

ksy90101 commented on src/main/java/gift/validator/ProductNameValidator.java:
```diff
@@ -0,0 +1,42 @@
+            return false;
+        }
+
+        if (!name.matches("[a-zA-Z0-9가-힣\\(\\)\\[\\]\\-+&_\\/\\s]+")) {
+            context.disableDefaultConstraintViolation();
+            context.buildConstraintViolationWithTemplate("상품 이름에는 (), [], -, +, &, _, /, 공백을 제외한 특수 문자를 사용할 수 없습니다.")
+                    .addConstraintViolation();
+            return false;
+        }
```
이부분은 @Regex로 해결해볼수 있지 않을까요?

ksy90101 commented on src/main/java/gift/ProductController.java:
```diff
@@ -59,4 +59,12 @@ public ResponseEntity<String> handleProductAlreadyExistsException(ProductAlready
         return ResponseEntity.status(HttpStatus.CONFLICT).body(e.getMessage());
     }
 
+    @ExceptionHandler(value = MethodArgumentNotValidException.class)
+    public ResponseEntity<String> handleProductNameException(MethodArgumentNotValidException e) {
+        String errorMessage = e.getBindingResult().getFieldErrors().stream()
+                .map(error -> error.getDefaultMessage())
+                .findFirst()
+                .orElse("올바르지 않은 입력 방식입니다.");
+        return ResponseEntity.badRequest().body(errorMessage);
+    }
```
ExcpetionHandler를 따로 클래스로 분리해서 관리해보면 어덜까요?

ksy90101 reviewed (changes requested):
2단계 고생많으셨습니다~
한번 기본 Validate 어노테이션을 학습해서 리팩토링 해보시면 좋을거 같아요!
언제든 질문이 있다면 DM주세요!

pkyung committed:
Refactor: 기본 Validate 어노테이션으로 수정함

Validator를 제작해서 처리했던 상품 이름 validation을 기본 Validate 어노테이션을 활용하여 구현함

pkyung committed:
Refactor: ExceptionHandler 클래스로 분리

컨트롤러에 예외처리 로직이 있어서 코드가 복잡해졌다. 이를
ProductExceptionHandler 클래스를 만들어 ProductController의 예외를
처리하도록 구현했다.

ksy90101 commented on src/main/java/gift/ProductController.java:
```diff
@@ -1,26 +1,23 @@
 
     @PostMapping("/products")
-    public ResponseEntity makeProduct(@RequestBody ProductRequestDto requestDto) {
+    public ResponseEntity makeProduct(@RequestBody @Validated ProductRequestDto requestDto) {
```
Valid와 Validated의 차이를 알아보면 좋을거 같습니다.

https://velog.io/@tjdtn0219/%EC%8A%A4%ED%94%84%EB%A7%81Valid-vs-Validated

ksy90101 commented on src/main/java/gift/ProductExceptionHandler.java:
```diff
@@ -0,0 +1,32 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice(assignableTypes = ProductController.class)
+public class ProductExceptionHandler {
```
👍

ksy90101 commented on src/main/java/gift/ProductExceptionHandler.java:
```diff
@@ -0,0 +1,32 @@
+
+    @ExceptionHandler(ProductNotFoundException.class)
+    public ResponseEntity<String> handleProductNotFoundException(ProductNotFoundException e) {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
```
```suggestion
        return ResponseEntity.notFound().body(e.getMessage());
```
이렇게 적을수 있습니다~

pkyung replied:
notfound가 body를 갖고 있지 않는 것 같습니다.

ksy90101 replied:
앗 죄송해요 ㅠㅠ 다른 코드랑 헷갈렸네요.
지금처럼 작성하시면 됩니다!

ksy90101 reviewed (changes requested):
간단한 수정정도만 작성해봤습니다! 궁금한거 있으면 언제든 DM주세요!

pkyung committed:
Refactor: Validated를 Valid로 변경

ksy90101 reviewed (approved):
고생많으셨어요~ 다음 미션도 홧팅입니다!
언제든 궁금한게 있다면 디엠주세요!
